### PR TITLE
Keymaster CTS Fixes

### PR DESCRIPTION
**Pull request by subrahmanyaman**

### CODE REVIEW

**Review by mdwivedi — CHANGES_REQUESTED** [review]

File: Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java [hunk 1]
```diff
@@ -49,16 +51,25 @@ public void onConsolidate() {
...
-    }
-    repository.onRestore(element, dataBaseVersion, CURRENT_DATABASE_VERSION);
-    seProvider.onRestore(element, dataBaseVersion, CURRENT_DATABASE_VERSION);
-    if (dataBaseVersion == INVALID_DATA_VERSION) {
-      handleDataUpgradeToVersion1();
+    byte magicNumber = element.readByte();
+    if (magicNumber != KMKeymasterApplet.KM_MAGIC_NUMBER) {
+      // Previous version of the applet does not have versioning support.
+      // In this case the first byte is the provision status.
+      provisionStatus = magicNumber;
```
[the remark this answers is not in the document]
Comment (subrahmanyaman) [reply Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java:58]: As discussed in the meeting no validation is required.

File: Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java [hunk 2]
```diff
@@ -49,16 +51,25 @@ public void onConsolidate() {
...
+      keymasterState = element.readByte();
```
[the remark this answers is not in the document]
Comment (subrahmanyaman) [reply Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java:59]: As discussed in the meeting no validation is required.

File: Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java [hunk 3]
```diff
@@ -49,16 +51,25 @@ public void onConsolidate() {
...
+      repository.onRestore(element, packageVersion, CURRENT_PACKAGE_VERSION);
+      seProvider.onRestore(element, packageVersion, CURRENT_PACKAGE_VERSION);
+      handleDataUpgradeToVersion1_1();
+    } else {
+      byte[] version = (byte[]) element.readObject();
+      if (!isUpgradeAllowed(version)) {
+        ISOException.throwIt(ISO7816.SW_CONDITIONS_NOT_SATISFIED);
+      }
+      packageVersion = version;
+      provisionStatus = element.readByte();
```
[the remark this answers is not in the document]
Comment (subrahmanyaman) [reply Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java:69]: Optimized the code.

File: Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java [hunk 4]
```diff
@@ -49,16 +51,25 @@ public void onConsolidate() {
...
-    seProvider.onRestore(element, dataBaseVersion, CURRENT_DATABASE_VERSION);
-    if (dataBaseVersion == INVALID_DATA_VERSION) {
-      handleDataUpgradeToVersion1();
+    byte magicNumber = element.readByte();
+    if (magicNumber != KMKeymasterApplet.KM_MAGIC_NUMBER) {
+      // Previous version of the applet does not have versioning support.
+      // In this case the first byte is the provision status.
+      provisionStatus = magicNumber;
+      keymasterState = element.readByte();
+      repository.onRestore(element, packageVersion, CURRENT_PACKAGE_VERSION);
```
[the remark this answers is not in the document]
Comment (subrahmanyaman) [reply Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java:60]: Updated this code. In this scenario the packageVersion is null.
When applet is installed for the first time the package version is set to zeros and when applet is getting upgraded this packageVersion holds the previous package version. And for the old applets where no versioning information this value will be null.

File: Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java [hunk 5]
```diff
@@ -77,25 +88,53 @@ public Element onSave() {
...
     repository.onSave(element);
     seProvider.onSave(element);
     return element;
   }
 
   private short computePrimitveDataSize() {
-    // provisionStatus + keymasterState
-    return (short) 4;
+    // provisionStatus + keymasterState + magic byte
+    return (short) 3;
```
[the remark this answers is not in the document]
Comment (subrahmanyaman) [reply Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java:102]: The previous value used to be 2. I even verified the divegeek:Javacard_KM_41_AOSP_UPMerge branch the value is 2 only. May be in this pull request some how it is showing the diff between the previous commits and latest.

File: Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java [hunk 6]
```diff
@@ -77,25 +88,53 @@ public Element onSave() {
...
-    return (short) 0;
+    return (short) 1;
+  }
+  
+  public boolean isUpgradeAllowed(byte[] version) {
+    short oldMajorVersion = Util.getShort(version, (short) 0);
+    short oldMinorVersion = Util.getShort(version, (short) 2);
+    short currentMajorVersion = Util.getShort(CURRENT_PACKAGE_VERSION, (short) 0);
+    short currentMinorVersion = Util.getShort(CURRENT_PACKAGE_VERSION, (short) 2);
+    // Downgrade of the Applet is not allowed.
```
[the remark this answers is not in the document]
Comment (subrahmanyaman) [reply Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEApplet.java:114]: Yes I used this logic and simplified the code.

File: Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEProvider.java [hunk 7]
```diff
@@ -1235,19 +1235,18 @@ public void onSave(Element element) {
...
     masterKey = KMAESKey.onRestore(element);
     attestationKey = KMECPrivateKey.onRestore(element);
     preSharedKey = KMHmacKey.onRestore(element);
-    if (oldVersion == INVALID_DATA_VERSION) {
-      handleDataUpgradeToVersion1();
-    } else if (oldVersion <= currentVersion) {
-      computedHmacKey = KMHmacKey.onRestore(element);
+    short oldMajorVersion = Util.getShort(oldVersion, (short) 0);
+    short oldMinorVersion = Util.getShort(oldVersion, (short) 2);
+    if (oldMajorVersion == 0 && oldMinorVersion == 0) {
```
[the remark this answers is not in the document]
Comment (subrahmanyaman) [reply Applet/AndroidSEProvider/src/com/android/javacard/keymaster/KMAndroidSEProvider.java:1245]: As discussed in this link https://github.com/divegeek/JavaCardKeymaster/pull/69#discussion_r759680520. The value of the old applets package version will be null. so updated the code accordingly.

**Pull request merged**